evm/v0.5/contracts/dev/PrepaidAggregator.sol

```diff
@@ -0,0 +1,435 @@
+    private
+    onlyOnNewRound(_id)
+    onlyIfDelayed(_id)
+  {
+    currentRound = _id;
+    rounds[_id].details.maxAnswers = maxAnswerCount;
+    rounds[_id].details.minAnswers = minAnswerCount;
+    rounds[_id].details.paymentAmount = paymentAmount;
+
+    oracles[msg.sender].lastStartedRound = _id;
+
+    emit NewRound(_id, msg.sender);
+  }
+
+  function updateRoundAnswer(uint32 _id)
+    private
+    onlyIfMinAnswersReceived(_id)
+  {
+    int256 newAnswer = Median.calculate(rounds[_id].details.answers);
+    rounds[_id].answer = newAnswer;
+    rounds[_id].updatedTimestamp = block.number;
+    latestRound = _id;
+
+    emit AnswerUpdated(newAnswer, _id);
+  }
+
+  function payOracle(uint32 _id)
+    private
+  {
+    uint128 payment = rounds[_id].details.paymentAmount;
+    uint128 available = availableFunds.sub(payment);
+
+    availableFunds = available;
+    allocatedFunds = allocatedFunds.add(payment);
+    oracles[msg.sender].withdrawable = oracles[msg.sender].withdrawable.add(payment);
+
+    emit AvailableFundsUpdated(available);
+  }
+
+  function recordSubmission(int256 _answer, uint32 _id)
+    private
+    onlyIfAcceptingAnswers(_id)
+  {
+    rounds[_id].details.answers.push(_answer);
+    oracles[msg.sender].lastReportedRound = _id;
+    oracles[msg.sender].latestAnswer = _answer;
+  }
+
+  function deleteRound(uint32 _id)
+    private
+    onlyIfMaxAnswersReceived(_id)
+  {
+    delete rounds[_id].details;
+  }
+
+  /**
+   * Modifiers
+   */
+
+  modifier onlyValidOracleRound(uint32 _id) {
+    uint32 startingRound = oracles[msg.sender].startingRound;
+    require(startingRound != 0, "Only updatable by whitelisted oracles");
+    require(startingRound <= _id, "New oracles cannot participate in in-progress rounds");
+    require(oracles[msg.sender].endingRound >= _id, "Oracle has been removed from whitelist");
+    require(oracles[msg.sender].lastReportedRound < _id, "Cannot update round reports");
+    _;
+  }
+
+  modifier onlyIfMinAnswersReceived(uint32 _id) {
+    if (rounds[_id].details.answers.length >= rounds[_id].details.minAnswers) {
+      _;
+    }
+  }
+
+  modifier onlyIfMaxAnswersReceived(uint32 _id) {
+    if (rounds[_id].details.answers.length == rounds[_id].details.maxAnswers) {
+      _;
+    }
+  }
+
+  modifier onlyIfAcceptingAnswers(uint32 _id) {
+    require(rounds[_id].details.maxAnswers != 0, "Max responses reached for round");
+    _;
+  }
+
+  modifier onlyOnNewRound(uint32 _id) {
+    if (_id == currentRound.add(1)) {
+      _;
+    }
+  }
+
+  modifier onlyIfDelayed(uint32 _id) {
+    uint256 lastStarted = oracles[msg.sender].lastStartedRound;
+    if (_id > lastStarted + restartDelay) {
+      _;
+    }
+  }
+
+  modifier onlyValidRoundId(uint32 _id) {
+    require(_id == currentRound || _id == currentRound.add(1), "Must report on current round");
+    if (_id > 1) {
+      require(rounds[_id.sub(1)].updatedTimestamp > 0, "Cannot bump round until previous round has an answer");
```

Comment:
It might be safer to add a timeout, here, in case quorum can't be reached on the previous round for some reason.

Reply:
How about we allow the owner to start a new round?

Reply:
Sure, that works, too.